File: frontend/js/views/savings_waterfall.js
// Savings Waterfall — the financial order of operations. Allocate a
// month's available savings down the priority ladder (employer match,
// high-interest debt, emergency fund, tax-advantaged, then taxable) via
// /calc/savings-waterfall.

import { api } from '../api.js';
import { esc } from '../util.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';

const STEP_LABEL = {
    starter_emergency: 'Starter emergency buffer',
    employer_match: 'Capture employer match',
    high_interest_debt: 'High-interest debt payoff',
    full_emergency: 'Full emergency fund (3–6 mo)',
    tax_advantaged: 'Tax-advantaged (HSA / Roth)',
    max_retirement: 'Max retirement (401k / IRA)',
    taxable_brokerage: 'Taxable brokerage',
};

const FIELDS = [
    ['monthly_available', 'Monthly savings available ($)', 2000],
    ['starter_emergency_gap', 'Starter buffer remaining ($)', 1000],
    ['employer_match_monthly', 'Monthly to full employer match ($)', 500],
    ['high_interest_debt', 'High-interest debt balance ($)', 3000],
    ['full_emergency_gap', 'Full emergency fund remaining ($)', 12000],
    ['tax_advantaged_room_monthly', 'HSA + Roth room this month ($)', 1000],
    ['retirement_room_monthly', '401k/IRA room this month ($)', 1500],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });

export async function renderSavingsWaterfall(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.waterfall.h1.title">// SAVINGS WATERFALL</span></h1>
        <p class="muted small" data-i18n="view.waterfall.hint.intro">
            The financial order of operations: where the next dollar goes. Each month's
            savings flows down the ladder — secure the employer match (free money) and a
            starter buffer before chasing returns, kill high-interest debt before a taxable
            brokerage, fill tax-advantaged space (HSA, Roth) before ordinary investing —
            with whatever is left landing in a taxable account.
        </p>
        <div class="chart-panel">
            <h2 data-i18n="view.waterfall.h2.inputs">Your numbers</h2>
            <form id="wf-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.waterfall.label.${key}">${label}</span>
                        <input type="number" step="1" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
                <button class="primary" type="submit" data-i18n="view.waterfall.btn.plan">Build waterfall</button>
            </form>
        </div>
        <div id="wf-result"></div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#wf-form');
    form.addEventListener('submit', async (e) => {
        e.preventDefault();
        const fd = new FormData(e.target);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        try {
            const plan = await api.calcSavingsWaterfall(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, plan);
        } catch (err) {
            showToast(err.message || t('view.waterfall.toast.error'), { level: 'error' });
        }
    });
    // Auto-run once with the defaults so the ladder is visible on open.
    form.dispatchEvent(new Event('submit'));
}

function renderResult(mount, plan) {
    const el = mount.querySelector('#wf-result');
    const total = Number(plan.total_allocated);
    const rows = plan.allocations.map((a) => {
        const amt = Number(a.amount);
        const pct = total > 0 ? (amt / total) * 100 : 0;
        return `<tr>
            <td>${esc(STEP_LABEL[a.step] || a.step)}</td>
            <td>${money(amt)}</td>
            <td class="muted">${pct.toFixed(1)}%</td>
        </tr>`;
    }).join('');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.waterfall.h2.plan">This month's allocation</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.waterfall.card.allocated">Allocated</div>
                    <div class="value">${money(plan.total_allocated)}</div></div>
            </div>
            <table class="trades">
                <thead><tr>
                    <th data-i18n="view.waterfall.th.step">Priority</th>
                    <th data-i18n="view.waterfall.th.amount">Amount</th>
                    <th data-i18n="view.waterfall.th.share">Share</th>
                </tr></thead>
                <tbody>${rows || `<tr><td colspan="3" class="muted" data-i18n="view.waterfall.empty">Nothing to allocate.</td></tr>`}</tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/savings_waterfall.rs
//! Savings waterfall — the financial order of operations.
//!
//! Where should the next dollar of savings go? The widely-taught answer
//! is a priority ladder: secure the employer match (free money) and clear
//! a starter emergency buffer before chasing returns, kill high-interest
//! debt before investing taxable, fill tax-advantaged space (HSA, Roth)
//! before a brokerage. This allocates a month's available savings down
//! that ladder, each rung capped by its remaining gap, with whatever is
//! left landing in a taxable brokerage.
//!
//! Pure compute — every rung's gap is supplied by the caller, so the
//! ordering is the only opinion here.

use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Step {
    /// A small cash buffer (one month / a deductible) before anything else.
    StarterEmergency,
    /// Enough into the 401(k) to capture the full employer match.
    EmployerMatch,
    /// High-interest debt (cards, anything above ~the market's expected return).
    HighInterestDebt,
    /// The full 3–6 month emergency fund.
    FullEmergency,
    /// Tax-advantaged space with an edge: HSA (triple-advantaged), then Roth.
    TaxAdvantaged,
    /// Remaining tax-deferred retirement room (401k/IRA to the limit).
    MaxRetirement,
    /// Everything left over — an ordinary taxable brokerage (uncapped).
    TaxableBrokerage,
}

#[derive(Debug, Clone, Deserialize)]
pub struct WaterfallInput {
    /// Dollars available to allocate this month.
    pub monthly_available: Decimal,
    /// Remaining to reach the starter emergency buffer.
    pub starter_emergency_gap: Decimal,
    /// Monthly contribution needed to capture the full employer match.
    pub employer_match_monthly: Decimal,
    /// Balance of high-interest debt to retire.
    pub high_interest_debt: Decimal,
    /// Remaining to reach the full 3–6 month emergency fund.
    pub full_emergency_gap: Decimal,
    /// Remaining tax-advantaged room for the month (HSA + Roth).
    pub tax_advantaged_room_monthly: Decimal,
    /// Remaining tax-deferred retirement room for the month.
    pub retirement_room_monthly: Decimal,
}

#[derive(Debug, Clone, Serialize)]
pub struct Allocation {
    pub step: Step,
    pub amount: Decimal,
}

#[derive(Debug, Clone, Serialize)]
pub struct WaterfallPlan {
    pub allocations: Vec<Allocation>,
    pub total_allocated: Decimal,
    /// Always zero — the taxable brokerage absorbs any remainder — but
    /// returned for symmetry and to make a zero-budget case explicit.
    pub unallocated: Decimal,
}

/// Allocate `monthly_available` down the priority ladder, each rung
/// capped by its supplied gap; the remainder lands in a taxable
/// brokerage. A rung that gets nothing is omitted from the plan.
pub fn plan(input: &WaterfallInput) -> WaterfallPlan {
    let mut remaining = input.monthly_available.max(Decimal::ZERO);
    let rungs = [
        (input.starter_emergency_gap, Step::StarterEmergency),
        (input.employer_match_monthly, Step::EmployerMatch),
        (input.high_interest_debt, Step::HighInterestDebt),
        (input.full_emergency_gap, Step::FullEmergency),
        (input.tax_advantaged_room_monthly, Step::TaxAdvantaged),
        (input.retirement_room_monthly, Step::MaxRetirement),
    ];

    let mut allocations = Vec::new();
    for (gap, step) in rungs {
        let amount = remaining.min(gap.max(Decimal::ZERO));
        if amount > Decimal::ZERO {
            allocations.push(Allocation { step, amount });
            remaining -= amount;
        }
    }
    if remaining > Decimal::ZERO {
        allocations.push(Allocation { step: Step::TaxableBrokerage, amount: remaining });
        remaining = Decimal::ZERO;
    }

    let total_allocated = allocations.iter().map(|a| a.amount).sum();
    WaterfallPlan { allocations, total_allocated, unallocated: remaining }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn d(v: i64) -> Decimal {
        Decimal::from(v)
    }

    fn input() -> WaterfallInput {
        WaterfallInput {
            monthly_available: Decimal::ZERO,
            starter_emergency_gap: d(1_000),
            employer_match_monthly: d(500),
            high_interest_debt: d(3_000),
            full_emergency_gap: d(12_000),
            tax_advantaged_room_monthly: d(1_000),
            retirement_room_monthly: d(1_500),
        }
    }

    fn amount_for(plan: &WaterfallPlan, step: Step) -> Decimal {
        plan.allocations
            .iter()
            .find(|a| a.step == step)
            .map(|a| a.amount)
            .unwrap_or(Decimal::ZERO)
    }

    #[test]
    fn fills_rungs_in_priority_order() {
        // $2,000 this month: $1,000 starter, then $500 match, then $500
        // toward the $3k debt. Nothing reaches the later rungs.
        let p = plan(&WaterfallInput { monthly_available: d(2_000), ..input() });
        assert_eq!(amount_for(&p, Step::StarterEmergency), d(1_000));
        assert_eq!(amount_for(&p, Step::EmployerMatch), d(500));
        assert_eq!(amount_for(&p, Step::HighInterestDebt), d(500));
        assert_eq!(amount_for(&p, Step::FullEmergency), Decimal::ZERO);
        assert_eq!(p.total_allocated, d(2_000));
        assert_eq!(p.unallocated, Decimal::ZERO);
    }

    #[test]
    fn match_comes_before_debt() {
        // The whole point of the ordering: even with debt outstanding, a
        // dollar past the starter buffer captures the match first.
        let p = plan(&WaterfallInput { monthly_available: d(1_200), ..input() });
        assert_eq!(amount_for(&p, Step::StarterEmergency), d(1_000));
        assert_eq!(amount_for(&p, Step::EmployerMatch), d(200));
        assert_eq!(amount_for(&p, Step::HighInterestDebt), Decimal::ZERO);
    }

    #[test]
    fn overflow_lands_in_taxable() {
        // More than every gap combined: 1000+500+3000+12000+1000+1500 =
        // 19,000 of capped rungs; $25k leaves $6k for the brokerage.
        let p = plan(&WaterfallInput { monthly_available: d(25_000), ..input() });
        assert_eq!(amount_for(&p, Step::TaxableBrokerage), d(6_000));
        assert_eq!(p.total_allocated, d(25_000));
    }

    #[test]
    fn zero_budget_allocates_nothing() {
        let p = plan(&WaterfallInput { monthly_available: Decimal::ZERO, ..input() });
        assert!(p.allocations.is_empty());
        assert_eq!(p.total_allocated, Decimal::ZERO);
        assert_eq!(p.unallocated, Decimal::ZERO);
    }

    #[test]
    fn negative_budget_clamped_to_zero() {
        let p = plan(&WaterfallInput { monthly_available: d(-500), ..input() });
        assert!(p.allocations.is_empty());
        assert_eq!(p.total_allocated, Decimal::ZERO);
    }

    #[test]
    fn skips_already_satisfied_rungs() {
        // No starter gap and no match needed → the first dollars go
        // straight to debt.
        let p = plan(&WaterfallInput {
            monthly_available: d(1_000),
            starter_emergency_gap: Decimal::ZERO,
            employer_match_monthly: Decimal::ZERO,
            ..input()
        });
        assert_eq!(amount_for(&p, Step::HighInterestDebt), d(1_000));
        assert!(p.allocations.iter().all(|a| a.step != Step::StarterEmergency));
    }

    #[test]
    fn total_allocated_equals_available_when_capacity_exists() {
        for budget in [100i64, 1_500, 9_999, 18_999] {
            let p = plan(&WaterfallInput { monthly_available: d(budget), ..input() });
            assert_eq!(p.total_allocated, d(budget), "budget {budget}");
        }
    }
}
